### PR TITLE
Lazily decode SourceFile from metadata

### PR DESCRIPTION
**Pull request by cjgillot**
Currently, source files from foreign crates are decoded up-front from metadata.
Spans from those crates were matched with the corresponding source using binary search among those files.

This PR changes the strategy by matching spans to files during encoding. This allows to decode source files on-demand, instead of up-front. The on-disk format for spans becomes: `<tag> <position from start of file> <length> <file index> <crate (if foreign file)>`.

### CODE REVIEW

**Comment by cjgillot** [comment]
r? compiler

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 1]
```diff
@@ -66,13 +65,10 @@ pub(super) struct EncodeContext<'a, 'tcx> {
...
-    // `SourceFile` into the `GrowableBitSet`.
-    //
-    // This needs to be a `GrowableBitSet` and not a
-    // regular `BitSet` because we may actually import new `SourceFiles`
-    // during metadata encoding, due to executing a query
-    // with a result containing a foreign `Span`.
-    required_source_files: Option<GrowableBitSet<usize>>,
+    // `SourceFile` into the `FxIndexSet`.
+    // The order inside the `FxIndexSet` is used as on-disk
+    // order of `SourceFiles`, and encoded inside `Span`s.
```
Comment (Mark-Simulacrum) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:70]: Is it worth making this order sensitive like that? I could imagine some kind of poor effects on deterministic metadata contents -- maybe we can emit in order or even sorted by (remapped) file path?

File: compiler/rustc_span/src/source_map.rs [hunk 2]
```diff
@@ -336,7 +336,7 @@ impl SourceMap {
         mut file_local_non_narrow_chars: Vec<NonNarrowChar>,
         mut file_local_normalized_pos: Vec<NormalizedPos>,
         original_start_pos: BytePos,
-        original_end_pos: BytePos,
+        _original_end_pos: BytePos,
```
Comment (Mark-Simulacrum) [inline compiler/rustc_span/src/source_map.rs:339]: Should we delete this argument completely? I could see us asserting start to end is source_len, I guess.

File: compiler/rustc_metadata/src/rmeta/encoder.rs (hunk 1 above)
Comment (cjgillot) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:70]: The order of the `SourceFile`s here should be exactly the order in which spans are encoded. Therefore, a change in this order can only be caused by a change in encoded spans. As we can't really know which `SourceFile`s we will need in the end, I don't see how we can pre-sort them.

**Comment by klensy** [comment]
This changes encoding for `Span` in rmeta, shouldn't encoding in `rustc_query_impl` be changed too?

**Comment by cjgillot** [comment]
@klensy no. metadata and incremental cache have different formats, on purpose.
We can rely on the order given in the metadata file for the program behaviour. Everything will be recompiled when this order change.
On the other hand, the order in the incremental cache can change unpredictably between two compiler runs. (I mean: the compiler does not predict the programmers' modifications to the code.) So this format needs to recognize stuff whose order may have changed.
Does it answer your question?

**Pull request merged**